src/compiler/checker.ts

```diff
@@ -8131,6 +8131,9 @@ namespace ts {
                     return anyType;
                 }
             }
+            if (indexType.flags & TypeFlags.Never) {
+                return neverType;
+            }

```

Comment:
Move these three lines up and insert then before line 8122. That way an `obj[key]` expression where `key` is an expression of type `never` will produce `never` (instead of the `any` it produces now).